yt/frontends/gamer/cfields.pyx

```diff
@@ -0,0 +1,215 @@
+# distutils: include_dirs = LIB_DIR
+# distutils: libraries = STD_LIBS
+cimport cython
+cimport libc.math as math
+cimport numpy as np
+import numpy as np
+
+np.import_array()
+
+cdef np.float64_t h_eos4(np.float64_t e):
+    cdef np.float64_t x
+    x = 2.25 * e * e
+    return 2.5 * e + x / (1.0 + math.sqrt(x + 1.0))
+
+cdef np.float64_t _lorentz_factor_eos4(
+        np.float64_t rho, 
+        np.float64_t mx,
+        np.float64_t my,
+        np.float64_t mz,
+        np.float64_t e,
+        np.float64_t c
+    ):
+    cdef np.float64_t u2, c2, vx, vy, vz
+    cdef np.float64_t htilde, fac;
+
+    c2 = c * c
+    fac = (1.0 / (rho * (h_eos4(e) + 1.0))) ** 2
+    u2 = (mx * mx + my * my + mz * mz) * fac
+    return math.sqrt(1.0 + u2 / c2)
+
+def gamma_eos4(temp):
+    cdef np.float64_t[:] kT = temp.ravel()
+    
+    out = np.empty_like(kT)
+    cdef np.float64_t[:] outp = out.ravel()
+    cdef np.float64_t x, c_p, c_v
```

Comment:
If these `ravel` calls make copies of the arrays, we might want to use some other trick to force it to appear as flat memoryviews

Reply:
I think it doesn't. I tried to check:

```
          Memory usage: ▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▅███ (max:   2.12GB, growth rate:  77%)           
                       memcheck.py: % of time = 100.00% out of  64.53s.                       
       ╷       ╷       ╷       ╷        ╷       ╷               ╷       ╷                     
  Line │Time   │–––––– │–––––– │Memory  │–––––– │–––––––––––    │Copy   │                     
       │Python │native │system │Python  │net    │timeline/%     │(MB/s) │memcheck.py          
╺━━━━━━┿━━━━━━━┿━━━━━━━┿━━━━━━━┿━━━━━━━━┿━━━━━━━┿━━━━━━━━━━━━━━━┿━━━━━━━┿━━━━━━━━━━━━━━━━━━━━╸
     1 │       │    1% │       │        │    7M │▁▁▁▁           │       │import numpy as np   
     2 │    3% │       │       │  10%   │  119M │▁▁▁▁▁▁▁▁▁  10% │     2 │from yt.frontends.g  
     3 │       │       │       │        │       │               │       │                     
     4 │       │       │       │        │ 1.00G │▁▂  44%        │       │inp = np.random.ran  
     5 │       │       │       │        │       │               │       │                     
     6 │       │       │       │        │       │               │       │print(inp[0,...], i  
     7 │       │       │       │        │       │               │       │                     
     8 │       │       │       │        │       │               │       │fgen = SRHDFields(4  
     9 │    4% │       │  90%  │        │ 1.00G │▄▄▄  44%       │       │out = fgen.sound_sp  
    10 │       │       │       │        │       │               │       │print(out[0,...])    
    11 │       │       │       │        │       │               │       │                     
       ╵       ╵       ╵       ╵        ╵       ╵               ╵       ╵                     
Top net memory consumption, by line:
(1)     4:  1025 MB                                                                            
(2)     9:  1024 MB                                                                            
(3)     2:   119 MB                                                                            
(4)     1:     7 MB    
```
also inserted `sleep()` in there and manually verified by top.
I think it's also indicated by how `out` and `outp` works. I do:

```
out = np.empty_like(dens)
cdef np.float64_t[:] outp = out.ravel()

for i in range(outp.shape[0]):
    outp[i] = ...
return out
```
note that I'm assigning values to a memoryview of a `.ravel()` and immediately afterwards I just return `out`. There's no copying I do by hand and it "just works".